common/transport/amqp/src/amqp.ts

```diff
@@ -578,33 +578,63 @@ export class Amqp {
 
-            detachLink(this._cbs, () => {
-              let remainingLinks = [];
-              for (let senderEndpoint in this._senders) {
-                if (this._senders.hasOwnProperty(senderEndpoint)) {
-                  remainingLinks.push(this._senders[senderEndpoint]);
-                  delete this._senders[senderEndpoint];
-                }
+            let remainingLinks = [];
+            for (let senderEndpoint in this._senders) {
+              if (this._senders.hasOwnProperty(senderEndpoint)) {
+                remainingLinks.push(this._senders[senderEndpoint]);
+                delete this._senders[senderEndpoint];
               }
+            }
 
-              for (let receiverEndpoint in this._receivers) {
-                if (this._receivers.hasOwnProperty(receiverEndpoint)) {
-                  remainingLinks.push(this._receivers[receiverEndpoint]);
-                    delete this._receivers[receiverEndpoint];
-                }
+            for (let receiverEndpoint in this._receivers) {
+              if (this._receivers.hasOwnProperty(receiverEndpoint)) {
+                remainingLinks.push(this._receivers[receiverEndpoint]);
+                  delete this._receivers[receiverEndpoint];
               }
+            }
 
-              /*Codes_SRS_NODE_COMMON_AMQP_16_034: [The `disconnect` method shall detach all open links before disconnecting the underlying AMQP client.]*/
-              async.each(remainingLinks, detachLink, () => {
-                sessionEnd((sessionError) => {
-                  disconnect((disconnectError) => {
-                    manageConnectionHandlers('removeListener');
-                    const finalError = err || sessionError || disconnectError;
-                    this._fsm.transition('disconnected', disconnectCallback, finalError);
+            //
+            // Depending on the mode of failure it is possible that no network activity is occurring.
+            // However, the SDK *might* not have noticed that yet.
+            //
+            // If this happens then rundown code will simply hang waiting for a response which will never
+            // occur.
+            //
+            // To deal with this, we encapsulate the rundown code in a function and we invoke it.
+            //
+            // First, however, we will spin off a timer to execute the very same rundown code in 45 seconds, but that
+            // function invocation will use API that do NOT expect a reply.
+            //
+            // If the initial function invocation actually doesn't hang due to lack of communication, it's
+            // last action will be to clear the timeout and consequently the timeout invocation call to the
+            // rundown code will NOT occur.
+            //
+            let rerunWithForceTimer: any = undefined;
+            const rundownConnection = (force: boolean) => {
+              if (force) {
+                this._disconnectionOccurred = true;
+              }
```

Comment:
nit: this._disconnectionOccurred = force :)

Reply:
Actually, after I left this afternoon I recalled about why I wrote it this way.

this._disconnectionOccurred might already be true when we enter this function.

Thus if we simply set it to the value of force we, on  the first call (the call NOT in the setTimer), will stomp a true to a false.

Not what we want.

In a sense this is why the force parameter name is good.

I will add a comment to call this out.

---
In reply to: [261444657](https://github.com/Azure/azure-iot-sdk-node/pull/499#discussion_r261444657) [](ancestors = 261444657)